### DISCUSSION TITLE
Estimation not proceeding from initial estimates, tips/tricks?

### DISCUSSION THREAD

**Discussion by deleveld** [nlmixr2/nlmixr2#257] [General]
Hi All,

I am new to nlmixr2 and have been trying to compare its estimation results to other software using a previous analysis. Info about the analysis can be found here: https://www.page-meeting.org/?abstract=834

What I am noticing is that for some data sets estimation wont proceed from the initial estimates and the final estimates are the same as the initial. NONMEM on the same dataset will converge. Is this a known thing? Any tips and tricks?

With NONMEM if convergence is poor I have found that just a few FO iterations will allow a subsequent FOCE estimation to converge. Is there a way to do something like this in nlmixr2?

warm regards,

Douglas Eleveld

**Comment by mattfidler** [comment]
>  Is this a known thing? Any tips and tricks?

Every dataset is different, so this is not a known thing.  However, I know that NONMEM and nlmixr2 use different algorithms for optimizing the likelihood.  They use a proprietary closed source optimization algorithm (so that is different). 

However, if the optimization technique in nlmixr2 for focei isn't working, you can change the optimization algorithm and sometimes it gives different/better results.  You can also switch between different methods and combine them:

``` r
library(nlmixr2)
#> Loading required package: nlmixr2data

one.compartment <- function() {
  ini({
    tka <- 0.45; label("Ka")
    tcl <- 1; label("Cl")
    tv <- 3.45; label("V")
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd <- 0.7
  })
  # and a model block with the error specification and model specification
  model({
    ka <- exp(tka + eta.ka)
    cl <- exp(tcl + eta.cl)
    v <- exp(tv + eta.v)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp ~ add(add.sd)
  })
}

# Default options
fit <- nlmixr2(one.compartment, theo_sd,  est="focei", foceiControl(print=0))
#> → loading into symengine environment...
#> → pruning branches (`if`/`else`) of full model...
#> ✔ done
#> → calculate jacobian
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → calculate sensitivities
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → calculate ∂(f)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → calculate ∂(R²)/∂(η)
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in inner model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in EBE model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → compiling inner model...
#> using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
#> ✔ done
#> → finding duplicate expressions in FD model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in FD model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → compiling EBE model...
#> using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
#> ✔ done
#> → compiling events FD model...
#> using C compiler: ‘gcc (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0’
#> ✔ done
#> rxode2 2.1.3.9000 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
#> calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 5952
#> → compress parHistData in nlmixr2 object, save 6560

# Can change the outer optimization method:
# For example bobyqa is a good gradient free optimization method
fit <- nlmixr2(one.compartment, theo_sd,  est="focei",
               foceiControl(print=0, outerOpt="bobyqa"))
#> calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 5952
#> → compress parHistData in nlmixr2 object, save 10336


# You can pipe together two different optimization methods, for
# example: This changes the inner optimization (finding the ETAs) and
# outer optimization steps:
fit <- nlmixr2(one.compartment, theo_sd,  est="focei",
               foceiControl(print=0, outerOpt="bobyqa", innerOpt="n1qn1")) %>%
  nlmixr2(est="focei", control=foceiControl(print=0, outerOpt="nlminb",
                                            innerOpt="BFGS"))
#> calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 5952
#> → compress parHistData in nlmixr2 object, save 10336
#> ℹ use `data` from prior/supplied fit
#> ℹ use `est` from prior/supplied fit
#> ℹ use/adapt `control` from prior/supplied fit
#> calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00 
#> done
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 5952
#> → compress parHistData in nlmixr2 object, save 8744
```

<sup>Created on 2024-08-15 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>


If it the problem is with linear solved systems, sometimes they are mis-behaving, so one of the many things on my list is to change the linear compartment solutions to those used in the R package `wnl` (instead of the advan solutions).

Another possible tip is , since one of the estimation methods is NONMEM you can check the translation and model differences by importing from NONMEM using `nonmem2rx()`  This will ensure the data/model are giving similar results.  If they are not this is a good place to start for debugging such a project.

> With NONMEM if convergence is poor I have found that just a few FO iterations will allow a subsequent FOCE estimation to converge. Is there a way to do something like this in nlmixr2?

Yes. However, `fo` is not often used and is not as well tested (and it seems that you cannot use the maximum number of iterations in `fo`.  For these reasons, I have a `as.function()` in part of the pipe, which does the same as the piping does in general, reads the population estimates and then uses them in the next estimation:


```r
library(nlmixr2)

# You can pipe together two different optimization methods, for
# example: This changes the inner optimization (finding the ETAs) and
# outer optimization steps:
fit <- nlmixr2(one.compartment, theo_sd,  est="fo", list(print=0)) %>%
  as.function() %>%
  nlmixr2(., theo_sd, est="focei", control=foceiControl(print=0, outerOpt="nlminb",
                                            innerOpt="BFGS"))
```

**Reply by deleveld** [reply]
Hi Matt,

Using different outerOpt and innerOpt doesnt seem to change very much. 
The model is just a $PRED like model. 
Running the NONMEM model through nonmem2rx seems to give essentially the same model code. 

If you are interested Ive attached the dataset, model code and the results. 
sink doesnt seem to capture all output, but it is still hopefully interpretable.
[
[sim.3.txt](https://github.com/user-attachments/files/16641156/sim.3.txt)
[control.3.R.txt](https://github.com/user-attachments/files/16641157/control.3.R.txt)
[control.3.R.out.txt](https://github.com/user-attachments/files/16641158/control.3.R.out.txt)
](url)
warm regards,

Douglas Eleveld
